File: 2025/day05/cpp/day05.cpp
#include <fstream>
#include <iostream>
#include <string>
#include <filesystem>
#include <vector>
#include <algorithm>

using namespace std;

using ll = long long;
using range = pair<ll, ll>;


void parse(const string& filename, vector<range>& ranges, vector<ll>& ingredients)
{
	if (!filesystem::exists(filename))
	{
		string fileInfo;
		fileInfo += "This is the current working directory: " + filesystem::current_path().string() + ".\n";
		fileInfo += "Filename supplied: " + filename + ".\n";
		fileInfo += "Resolves to absolute path: " + filesystem::absolute(filename).string() + ".\n";

		cerr << fileInfo << endl;
		cerr << "Error: file does not exist: " << filesystem::absolute(filename) << endl;
		throw runtime_error("file does not exist");
	}

	ifstream f(filename);
	if (!f.is_open())
	{
		cerr << "Error opening file: " << filename << endl;
		throw runtime_error("error opening file");
	}

	bool first_part = true;
	string line;

	while (getline(f, line))
	{
		if (line.empty())
		{
			first_part = false;
			continue;
		}

		if (first_part)
		{
			// ranges
			size_t dash_pos = line.find('-');
			ll start = stoll(line.substr(0, dash_pos));
			ll end = stoll(line.substr(dash_pos + 1));
			ranges.emplace_back(start, end);
		}
		else
		{
			// ingredients
			ingredients.emplace_back(stoll(line));
		}
	}

	f.close();
}

vector<range> merge_ranges(vector<range>& ranges)
{
	// sort input on start
	sort(ranges.begin(), ranges.end());

	vector<range> ans;

	for (const auto& range : ranges)
	{
		if (ans.empty())
		{
			ans.emplace_back(range);
		}
		else
		{
			ll prev_end = ans.back().second;
			ll curr_start = range.first;
			ll curr_end = range.second;

			// start new range
			if (curr_start > prev_end)
			{
				ans.emplace_back(range);
			}
			// extend prev range
			else
			{
				if (curr_end > prev_end)
				{
					ll start = ans.back().first;
					ans.back() = {start, curr_end};
				}
			}
		}
	}

	return ans;
}

void part1(const vector<range>& ranges, const vector<ll>& ingredients)
{
	ll fresh_cnt = 0;

	for (ll ingredient : ingredients)
	{
		for (const auto& [start, end] : ranges)
		{
			if (start <= ingredient && ingredient <= end)
			{
				fresh_cnt++;
				break;
			}
		}
	}

	cout << "Part 1: " << fresh_cnt << endl;
}

void part2(vector<range>& ranges)
{
	auto merged = merge_ranges(ranges);

	ll total_span = 0;
	for (const auto& [start, end] : merged)
	{
		total_span += end - start + 1;
	}

	cout << "Part 2: " << total_span << endl;
}

int main()
{
	//string filename = "../../../../example";
	string filename = "../../../../input";

	vector<range> ranges;
	vector<ll> ingredients;

	parse(filename, ranges, ingredients);
	part1(ranges, ingredients);
	part2(ranges);

	return 0;
}

File: 2025/day05/day05.py
from dataclasses import dataclass
import time
import math
import re
import copy
from collections import Counter
import sys
import os
from collections import defaultdict
import itertools


def parse(filename) -> tuple[list[tuple[int, int]], list[int]]:
    ranges = []
    ingredients = []
    with open(filename) as f:
        first_part = True
        while line := f.readline():
            line = line.strip()

            if len(line) == 0:
                first_part = False
                continue

            if first_part:
                # ranges
                start, end = line.split("-")
                ranges.append((int(start), int(end)))
            else:
                # ingredients
                ingredients.append(int(line))
    return ranges, ingredients


def part1(ranges: list[tuple[int, int]], ingredients: list[int]):
    fresh_cnt = 0

    for ingredient in ingredients:
        for start, end in ranges:
            if start <= ingredient <= end:
                fresh_cnt += 1
                break

    print("Part 1:", fresh_cnt)


def merge(intervals: list[tuple[int, int]]) -> list[tuple[int, int]]:
    # sort input on start
    intervals.sort()

    ans = []

    for interval in intervals:
        if len(ans) == 0:
            ans.append(interval)
        else:
            prev_end = ans[-1][1]

            curr_start = interval[0]
            curr_end = interval[1]

            # start new interval
            if curr_start > prev_end:
                ans.append(interval)

            # extend prev interval
            else:
                if curr_end > prev_end:
                    last_interval = ans[-1]
                    start, _ = last_interval
                    ans[-1] = (start, curr_end)

    return ans


def part2(ranges: list[tuple[int, int]]):
    ranges = merge(ranges)

    total_span = 0
    for start, end in ranges:
        total_span += end - start + 1

    print("Part 2:", total_span)


filename = "day05/example"
filename = "day05/input"

ranges, ingredients = parse(filename)

part1(ranges, ingredients)
part2(ranges)
